### PR TITLE
Rework rgb for fun vehicle and paintgun

### PR DESCRIPTION
**Pull request by TheGreenBandit**
uses different rainbow loops for each so no overlapping speed.

### CODE REVIEW

**Comment by BugisoftRSG** [comment]
Wouldn't it be better to have a future proof and updated version of rgb controller so other features like rainbow traffic may use it instead of limiting it to rainbow_paint_gun?

**Comment by TheGreenBandit** [comment]
> Wouldn't it be better to have a future proof and updated version of rgb controller so other features like rainbow traffic may use it instead of limiting it to rainbow_paint_gun?

that was my original plan with the original pr but many are having problems with the overlapping of the speeds so i figured putting different rainbow loops would be easier as it could just be copied over to other spots

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 1]
```diff
@@ -3,34 +3,76 @@
...
+				{
+					blue += g.vehicle.rainbow_paint.speed;
+					green -= g.vehicle.rainbow_paint.speed;
+				}
+				if (blue > 0 && green == 0)
+				{
+					red += g.vehicle.rainbow_paint.speed;
+					blue -= g.vehicle.rainbow_paint.speed;
+				}
+				if (red > 255)
```
Comment (xiaoxiao921) [inline src/backend/looped/vehicle/rgb_paint.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 2]
```diff
@@ -3,34 +3,76 @@
...
+				{
+					red = 255;
+				} //checks
+				if (green > 255)
+				{
+					green = 255;
```
Comment (xiaoxiao921) [inline src/backend/looped/vehicle/rgb_paint.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 3]
```diff
@@ -3,34 +3,76 @@
...
+				}
+				if (blue > 255)
+				{
+					blue = 255;
+				}
+				if (red < 0)
+				{
+					red = 0;
```
Comment (xiaoxiao921) [inline src/backend/looped/vehicle/rgb_paint.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 4]
```diff
@@ -3,34 +3,76 @@
...
+				{
+					red = 255;
+				} //checks
+				if (green > 255)
+				{
+					green = 255;
+				}
+				if (blue > 255)
+				{
+					blue = 255;
```
Comment (xiaoxiao921) [inline src/backend/looped/vehicle/rgb_paint.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 5]
```diff
@@ -3,34 +3,76 @@
...
+				}
+				if (red < 0)
+				{
+					red = 0;
+				}
+				if (green < 0)
+				{
+					green = 0;
```
Comment (xiaoxiao921) [inline src/backend/looped/vehicle/rgb_paint.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/vehicle/rgb_paint.cpp [hunk 6]
```diff
@@ -3,34 +3,76 @@
...
+				}
+				if (blue < 0)
+				{
+					blue = 0;
```
Comment (xiaoxiao921) [inline src/backend/looped/vehicle/rgb_paint.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/weapons/paintgun.cpp [hunk 7]
```diff
@@ -2,19 +2,44 @@
...
+				green -= g.weapons.paintgun.speed;
+			}
+			if (blue > 0 && green == 0)
+			{
+				red += g.weapons.paintgun.speed;
+				blue -= g.weapons.paintgun.speed;
+			}
+			if (red > 255) {red = 255;} //checks to prevent freezing
+			if (green > 255) {green = 255;}
+			if (blue > 255) {blue = 255;}
```
Comment (xiaoxiao921) [inline src/backend/looped/weapons/paintgun.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/weapons/paintgun.cpp [hunk 8]
```diff
@@ -2,19 +2,44 @@
...
+				blue += g.weapons.paintgun.speed;
+				green -= g.weapons.paintgun.speed;
+			}
+			if (blue > 0 && green == 0)
+			{
+				red += g.weapons.paintgun.speed;
+				blue -= g.weapons.paintgun.speed;
+			}
+			if (red > 255) {red = 255;} //checks to prevent freezing
+			if (green > 255) {green = 255;}
```
Comment (xiaoxiao921) [inline src/backend/looped/weapons/paintgun.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

File: src/backend/looped/weapons/paintgun.cpp [hunk 9]
```diff
@@ -2,19 +2,44 @@
...
+			{
+				blue += g.weapons.paintgun.speed;
+				green -= g.weapons.paintgun.speed;
+			}
+			if (blue > 0 && green == 0)
+			{
+				red += g.weapons.paintgun.speed;
+				blue -= g.weapons.paintgun.speed;
+			}
+			if (red > 255) {red = 255;} //checks to prevent freezing
```
Comment (xiaoxiao921) [inline src/backend/looped/weapons/paintgun.cpp]: https://en.cppreference.com/w/cpp/algorithm/clamp

**Comment by xiaoxiao921** [comment]
Actually use https://en.cppreference.com/w/cpp/algorithm/clamp since you both do a min and max check

**Review by Yimura — APPROVED** [review]
Works as expected.

**Pull request merged**